### PR TITLE
feat(mithril): epic7 pre-investigation: Scaffolding markdown parser

### PR DESCRIPTION
**Pull request by igarashitm**
https://github.com/user-attachments/assets/b277172e-c0be-44ea-ba4a-13df02ab9810

## Summary by CodeRabbit

- **New Features**
  - Enabled full parsing and serialization of numbered custom instruction steps (titles + markdown content) for custom modes.
  - Custom instruction blocks now automatically include a standard system-instruction preamble and a hard-rules trailer.
  - Updated visual mode rendering to show step titles and include instruction content in the node descriptions.
  - Custom mode templates now generate an initial `customInstructions` section for new modes.

- **Tests**
  - Added comprehensive parser/serializer and round-trip tests, including edge cases and formatting/indentation.

### CODE REVIEW

File: packages/ui/src/models/custom-mode/custom-instructions-parser.ts [hunk 1]
```diff
@@ -1,24 +1,123 @@
...
+ *    - sub-bullet               ← nested list items stay inside rawContent
+ *    - sub-bullet
+ * 2. Next step
+ *    ...
+ *
+ * > Hard rules                  ← blockquote — ignored for now
+ * ```
+ *
+ * TODO: Epic 7 — decide what to do with the preamble paragraph (system
+ * instructions text above the heading). Currently skipped entirely.
```
Comment (PVinaches) [inline packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: Could we add that automatically without the need to implement a new node? That is needed from our end for some tools to be detected by the LLM but the user shouldn't know about it more than the automatic addition of it in the mode

File: packages/ui/src/models/custom-mode/custom-instructions-parser.ts [hunk 2]
```diff
@@ -1,24 +1,123 @@
...
+ *
+ * TODO: Epic 7 — decide whether the blockquote "Hard rules" section should
+ * be exposed as metadata on the entity or as a special node type.
```
Comment (PVinaches) [inline packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: I would leave the hard rules as a text node for now. We will add manually the Hard rules in title and let's think about further improvements in a later phase

Comment (igarashitm) [reply packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: experiment: show `Hard Rules` as a custom mode property
Screenshot From 2026-07-15 15-13-22

File: packages/ui/src/models/custom-mode/custom-instructions-parser.ts (hunk 1 above)
Comment (igarashitm) [reply packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: done

File: packages/ui/src/models/custom-mode/custom-instructions-parser.ts (hunk 2 above)
Comment (PVinaches) [reply packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: I like this one too. @shivamG640 what do you think about adding this component to the catalog?

File: packages/ui/src/models/custom-mode/custom-instructions-parser.ts [hunk 3]
```diff
@@ -1,24 +1,186 @@
...
+      } satisfies CustomInstructionsNode;
+    });
+  }
+
+  /**
+   * Collects plain-text rules from all top-level blockquotes.
+   * Each unordered list item inside a blockquote becomes one rule string.
+   * Paragraph children (e.g. the "Hard rules" header line) are skipped.
+   *
+   * TODO: Epic 7 — decide whether to preserve the blockquote heading as a label.
```
Comment (PVinaches) [inline packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: if the new node isn't added I suggest to follow your suggestion. I think we should also include as headings html elements (h1, h2, h3) if it is supported by the library too

Comment (PVinaches) [reply packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: decided in meeting to be markdown in description

File: packages/ui/src/models/custom-mode/custom-instructions-parser.ts (hunk 2 above)
Comment (PVinaches) [reply packages/ui/src/models/custom-mode/custom-instructions-parser.ts]: decided in meeting to be like custom instructions

**Review by PVinaches — APPROVED** [review]
